### ISSUE TITLE
mini-batch with no duplicate speaker_id

### ISSUE THREAD

**Issue by ductuantruong** [wenet-e2e/wespeaker#195]
Hi Wespeaker team,

Thank you for sharing your amazing work. I am trying to use your work in my research. However, I am not familiar with the data sampler using your toolkit. Could I ask how to modify your data sampler so that in a mini-batch, each training sample has a different speaker_id (no duplicate speaker_id in a mini-batch)?

Thank you for your support!

**Comment by JiJiJiang** [comment]
![image](https://github.com/wenet-e2e/wespeaker/assets/23723621/5b8cef60-6303-44df-8cdf-edbcb8837db7)
I think one simple method is to re-define your collate_fn in the dataloader, and filter the training samples of the same spk_id with a batch. Although the batch_size could be a little smaller, the training process should be normal.
You can refer our similar usage of customized collate_fn in [ssl/dataset.py](https://github.com/wenet-e2e/wespeaker/blob/fee11093e6090c72306ea3f2cb01414045c050c7/wespeaker/ssl/dataset/dataset.py#L27). Hope this solve your problem.

**Comment by ductuantruong** [comment]
Thank you for your suggestion! I will try that.

**Comment by mmmmayi** [comment]
Then if I want to use triplet loss and a batch contains N speakers with M different utts for each speaker, can I modify your data sampler?
